### PR TITLE
[INFRA/CORE] Add core components according to spec

### PR DESCRIPTION
**Pull request by lhl2617**
# Summary

- Implement chapter 3 in https://www.overleaf.com/read/swjhzbcxjnzr
- Add a config package to define initial state/other constants.
- Add to Client Interface (and their structs thereof) a `ReceiveGameStateUpdate` method called by `SOMASServer:updateIslands`.
- Don't hold pointers to clients in `gameState` so that we can safely pass `gameState` to clients--made a separate `SOMASServer.clientMap` to store this info.

## Additional Information

- Changed function names as seen fit (e.g. `init_iigo` -> `runIIGO`)
- Changed round to `season`
- Take cost of living at _end of turn_
- Add `GoStringer` implementations for the two iotas.

## ⚠️ NOT DONE YET
- Actions -- see #42 

## Test Plan

- set `config.MaxTurns` to 2 (as per current diff) and run for two turns 
- Note that `go run .` shows TURN 3 in the end -- this is because these are _initial game states on the turn_--TURN 3 -> end of Turn 2.

- `go test ./...`
```
go test ./...
?       github.com/SOMAS2020/SOMAS2020  [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team1   [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team2   [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team3   [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team4   [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team5   [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team6   [no test files]
ok      github.com/SOMAS2020/SOMAS2020/internal/common  (cached)
?       github.com/SOMAS2020/SOMAS2020/internal/common/config   [no test files]
ok      github.com/SOMAS2020/SOMAS2020/internal/common/rules    (cached)
?       github.com/SOMAS2020/SOMAS2020/internal/common/shared   [no test files]
ok      github.com/SOMAS2020/SOMAS2020/internal/server  (cached)
?       github.com/SOMAS2020/SOMAS2020/pkg/testutils    [no test files]
```

- `go run .`
```2020/12/09 15:08:29 [SERVER]: TURN: 1, Season: 1
2020/12/09 15:08:29 [SERVER]: start updateIsland
2020/12/09 15:08:29 [Team4]: Received game state update: {1 1 map[Team1:{100 Alive 0} Team2:{100 Alive 0} Team3:{100 Alive 0} Team4:{100 Alive 0} Team5:{100 Alive 0} Team6:{100 Alive 0}]}
2020/12/09 15:08:29 [Team5]: Received game state update: {1 1 map[Team1:{100 Alive 0} Team2:{100 Alive 0} Team3:{100 Alive 0} Team4:{100 Alive 0} Team5:{100 Alive 0} Team6:{100 Alive 0}]}
2020/12/09 15:08:29 [Team6]: Received game state update: {1 1 map[Team1:{100 Alive 0} Team2:{100 Alive 0} Team3:{100 Alive 0} Team4:{100 Alive 0} Team5:{100 Alive 0} Team6:{100 Alive 0}]}
2020/12/09 15:08:29 [Team1]: Received game state update: {1 1 map[Team1:{100 Alive 0} Team2:{100 Alive 0} Team3:{100 Alive 0} Team4:{100 Alive 0} Team5:{100 Alive 0} Team6:{100 Alive 0}]}
2020/12/09 15:08:29 [Team2]: Received game state update: {1 1 map[Team1:{100 Alive 0} Team2:{100 Alive 0} Team3:{100 Alive 0} Team4:{100 Alive 0} Team5:{100 Alive 0} Team6:{100 Alive 0}]}
2020/12/09 15:08:29 [Team3]: Received game state update: {1 1 map[Team1:{100 Alive 0} Team2:{100 Alive 0} Team3:{100 Alive 0} Team4:{100 Alive 0} Team5:{100 Alive 0} Team6:{100 Alive 0}]}
2020/12/09 15:08:29 [SERVER]: finish updateIsland
2020/12/09 15:08:29 [SERVER]: start runOrgs
2020/12/09 15:08:29 [SERVER]: start runIITO
2020/12/09 15:08:29 [SERVER]: finish runIITO
2020/12/09 15:08:29 [SERVER]: start runIIFO
2020/12/09 15:08:29 [SERVER]: finish runIIFO
2020/12/09 15:08:29 [SERVER]: start runIITO
2020/12/09 15:08:29 [SERVER]: finish runIITO
2020/12/09 15:08:29 [SERVER]: finish runOrgs
2020/12/09 15:08:29 [SERVER]: start endOfTurn
2020/12/09 15:08:29 [SERVER]: start probeDisaster
2020/12/09 15:08:29 [SERVER]: finish probeDisaster
2020/12/09 15:08:29 [SERVER]: start deductCostOfLiving
2020/12/09 15:08:29 [SERVER]: finish deductCostOfLiving
2020/12/09 15:08:29 [SERVER]: start updateIslandLivingStatus
2020/12/09 15:08:29 [SERVER]: finish updateIslandLivingStatus
2020/12/09 15:08:29 [SERVER]: finish endOfTurn
2020/12/09 15:08:29 [SERVER]: finish runTurn
2020/12/09 15:08:29 [SERVER]: start runTurn
2020/12/09 15:08:29 [SERVER]: TURN: 2, Season: 1
2020/12/09 15:08:29 [SERVER]: start updateIsland
2020/12/09 15:08:29 [Team2]: Received game state update: {1 2 map[Team1:{90 Alive 0} Team2:{90 Alive 0} Team3:{90 Alive 0} Team4:{90 Alive 0} Team5:{90 Alive 0} Team6:{90 Alive 0}]}
2020/12/09 15:08:29 [Team3]: Received game state update: {1 2 map[Team1:{90 Alive 0} Team2:{90 Alive 0} Team3:{90 Alive 0} Team4:{90 Alive 0} Team5:{90 Alive 0} Team6:{90 Alive 0}]}
2020/12/09 15:08:29 [Team4]: Received game state update: {1 2 map[Team1:{90 Alive 0} Team2:{90 Alive 0} Team3:{90 Alive 0} Team4:{90 Alive 0} Team5:{90 Alive 0} Team6:{90 Alive 0}]}
2020/12/09 15:08:29 [Team5]: Received game state update: {1 2 map[Team1:{90 Alive 0} Team2:{90 Alive 0} Team3:{90 Alive 0} Team4:{90 Alive 0} Team5:{90 Alive 0} Team6:{90 Alive 0}]}
2020/12/09 15:08:29 [Team6]: Received game state update: {1 2 map[Team1:{90 Alive 0} Team2:{90 Alive 0} Team3:{90 Alive 0} Team4:{90 Alive 0} Team5:{90 Alive 0} Team6:{90 Alive 0}]}
2020/12/09 15:08:29 [Team1]: Received game state update: {1 2 map[Team1:{90 Alive 0} Team2:{90 Alive 0} Team3:{90 Alive 0} Team4:{90 Alive 0} Team5:{90 Alive 0} Team6:{90 Alive 0}]}
2020/12/09 15:08:29 [SERVER]: finish updateIsland
2020/12/09 15:08:29 [SERVER]: start runOrgs
2020/12/09 15:08:29 [SERVER]: start runIITO
2020/12/09 15:08:29 [SERVER]: finish runIITO
2020/12/09 15:08:29 [SERVER]: start runIIFO
2020/12/09 15:08:29 [SERVER]: finish runIIFO
2020/12/09 15:08:29 [SERVER]: start runIITO
2020/12/09 15:08:29 [SERVER]: finish runIITO
2020/12/09 15:08:29 [SERVER]: finish runOrgs
2020/12/09 15:08:29 [SERVER]: start endOfTurn
2020/12/09 15:08:29 [SERVER]: start probeDisaster
2020/12/09 15:08:29 [SERVER]: finish probeDisaster
2020/12/09 15:08:29 [SERVER]: start deductCostOfLiving
2020/12/09 15:08:29 [SERVER]: finish deductCostOfLiving
2020/12/09 15:08:29 [SERVER]: start updateIslandLivingStatus
2020/12/09 15:08:29 [SERVER]: finish updateIslandLivingStatus
2020/12/09 15:08:29 [SERVER]: finish endOfTurn
2020/12/09 15:08:29 [SERVER]: finish runTurn
2020/12/09 15:08:29 [SERVER]: Max turns '2' reached or exceeded
===== START OF TURN 1 (END OF TURN 0) =====
common.GameState{Season:0x1, Turn:0x1, ClientInfos:map[shared.ClientID]common.ClientInfo{Team1:common.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team2:common.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team3:common.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team4:common.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team5:common.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team6:common.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}}}
===== START OF TURN 2 (END OF TURN 1) =====
common.GameState{Season:0x1, Turn:0x2, ClientInfos:map[shared.ClientID]common.ClientInfo{Team1:common.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team2:common.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team3:common.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team4:common.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team5:common.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team6:common.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}}}
===== START OF TURN 3 (END OF TURN 2) =====
common.GameState{Season:0x1, Turn:0x3, ClientInfos:map[shared.ClientID]common.ClientInfo{Team1:common.ClientInfo{Resources:80, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team2:common.ClientInfo{Resources:80, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team3:common.ClientInfo{Resources:80, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team4:common.ClientInfo{Resources:80, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team5:common.ClientInfo{Resources:80, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team6:common.ClientInfo{Resources:80, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}}}

```

### CODE REVIEW

**Comment by QFSW** [comment]
# ⚠️Warning! This diff contains changes in protected folders

**Comment by lhl2617** [comment]
Will be waiting for Yannis to comment with explicit approval for these changes:

1. Rename `round` to `season`
2. Deduct taxes/cost of living at end of turn.

**Comment by lhl2617** [comment]
me: make diffs small
also me:
image

i wanna get everything in first quick now before we have more code to consider.

File: internal/common/game.go
Comment (QFSW) [inline internal/common/game.go]: This should be an enum because having separate bools allows for invalid states (e.g. dead and non-critical)

File: internal/server/helper.go
Comment (QFSW) [inline internal/server/helper.go]: I'm not sure if turns left is really the way to go, i feel like having a critical turns counter makes more sense

File: internal/server/helper.go
Comment (QFSW) [inline internal/server/helper.go:52]: Do we want events for things like entering/leaving critical status that might be of use to the client?

File: internal/server/server.go
Comment (QFSW) [inline internal/server/server.go:50]: not sure what `for !s.gameOver` means

File: internal/server/turn.go
Comment (QFSW) [inline internal/server/turn.go:137]: Should the `1` not be the startTurn?

**Review by QFSW — COMMENTED** [review]
Some small comments

File: internal/common/game.go
Comment (lhl2617) [reply internal/common/game.go]: Yep, will change -- wanted to do it but the critical days left was a bit of an ass to do, but it should make sense

File: internal/server/helper.go
Comment (lhl2617) [reply internal/server/helper.go]: Cool

File: internal/server/helper.go
Comment (lhl2617) [reply internal/server/helper.go:52]: The client should track that itself

File: internal/server/server.go
Comment (lhl2617) [reply internal/server/server.go:50]: While the game is not over

File: internal/server/turn.go
Comment (lhl2617) [reply internal/server/turn.go:137]: ? Dont understand this

Comment (lhl2617) [reply internal/server/turn.go:137]: address by removing initialturn and season nums

**Review by QFSW — APPROVED** [review]

**Comment by lhl2617** [comment]
Thanks @QFSW !

**Pull request merged**
